overlord/devicestate/devicestate.go

```diff
@@ -1411,35 +1457,264 @@ func createRecoverySystemTasks(st *state.State, label string, snapSetupTasks []s
+		finalize.WaitFor(create)
+		// finalize needs to know the label too
+		finalize.Set("recovery-system-setup-task", create.ID())
+
+		ts.AddTask(finalize)
+	}
+
+	return ts, nil
 }
 
-func CreateRecoverySystem(st *state.State, label string) (*state.Change, error) {
+// CreateRecoverySystemOptions is the set of options that can be used with
+// CreateRecoverySystem.
+type CreateRecoverySystemOptions struct {
+	// ValidationSets is a list of validation sets to use when creating the new
+	// recovery system. If provided, all snaps used to create recovery system
+	// will follow the constraints imposed by the validation sets. If required
+	// snaps are not present on the system, and LocalSnapSideInfos is not
+	// provided, then the snaps will be downloaded. Note that if validation sets
+	// are provided, all snaps in the model must be constrained by a validation
+	// set.
+	ValidationSets []*asserts.ValidationSet
+
+	// LocalSnapSideInfos is an optional list of snaps that will be used to
+	// create the new recovery system. If provided, this list must contain any
+	// snap that is not already installed that will be needed by the new recovery
+	// system.
+	LocalSnapSideInfos []*snap.SideInfo
+
+	// LocalSnapPaths is a list of paths to snaps that corresponds to the list
+	// of snap.SideInfo structs in SideInfos. It must be the same length as
+	// LocalSnapSideInfos.
+	LocalSnapPaths []string
+
+	// TestSystem is set to true if the new recovery system should be verified
+	// by rebooting into the new system, prior to marking it as a valid recovery
+	// system. If false, the system will immediately be considered a valid
+	// recovery system.
+	TestSystem bool
+
+	// MarkCurrent is set to true if the new recovery system should be
+	// marked as the current system.
+	MarkCurrent bool
+}
+
+var ErrNoRecoverySystem = errors.New("recovery system does not exist")
+
+// CreateRecoverySystem creates a new recovery system with the given label. See
+// CreateRecoverySystemOptions for details on the options that can be provided.
+func CreateRecoverySystem(st *state.State, label string, opts CreateRecoverySystemOptions) (chg *state.Change, err error) {
 	var seeded bool
-	err := st.Get("seeded", &seeded)
+	err = st.Get("seeded", &seeded)
 	if err != nil && !errors.Is(err, state.ErrNoState) {
 		return nil, err
 	}
 	if !seeded {
 		return nil, fmt.Errorf("cannot create new recovery systems until fully seeded")
 	}
-	chg := st.NewChange("create-recovery-system", fmt.Sprintf("Create new recovery system with label %q", label))
-	ts, err := createRecoverySystemTasks(st, label, nil)
+
+	valsets := snapasserts.NewValidationSets()
+	for _, vs := range opts.ValidationSets {
+		if err := valsets.Add(vs); err != nil {
+			return nil, err
+		}
+	}
+
+	if err := valsets.Conflict(); err != nil {
+		return nil, err
+	}
+
+	revisions, err := valsets.Revisions()
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := findModel(st)
```

Comment:
Is it always the case that we are going to create a recovery system from the current model?